File: x1/peragrams/peragrams.pl
:- [kattio].

% Func to count amount of each letter
countLetterFreq(String, Counts) :-
    string_chars(String, Chars), % string -> ['s', 't', 'r', 'i', 'n', 'g']
    msort(Chars, SortedChars), % sort the list of chars alphabetically
    sort(SortedChars, UniqueChars), % sort the list and remove duplicates for all UNIQUE letters
    
    % Operations in line (ii) stored in line (i) and then stored in line (iii)
    findall(
        Count, 
        (member(Letter, UniqueChars), countChars(SortedChars, Letter, Count)),
        Counts
    ).

% helper func to count letter freq
countChars([], _, 0).
countChars([CharH | CharT], Letter, Counter) :-
    countChars(CharT, Letter, SubCount), 
    (CharH == Letter -> Counter is SubCount+1 ; Counter = SubCount).

% func to count amount of oddfreq letters
findOddFreq(String, CharCountList, OddFreqCount) :-
    countLetterFreq(String, CharCountList), % finds list of ALL letter counts
    findall(
        1,
        (member(LetterFreq, CharCountList), isOdd(LetterFreq)),
        OddList
    ),
    length(OddList, OddFreqCount).

% func to check if number is odd
isOdd(N) :- N mod 2 =:= 1.

% remove 1 letter from each odd letter if more than 1 odd letter freq
removedLetters(String, OddFreqCount, RemovedLetters) :-
    findOddFreq(String, CharCount, OddFreqCount),
    (OddFreqCount > 1 -> RemovedLetters is OddFreqCount-1 ; RemovedLetters = 0).

% main func
main :- 
    read_string(Peragram),
    removedLetters(Peragram, OddFreqCount, RemovedLetters), 
    write(RemovedLetters), nl.

File: x1/peragrams/peragrams.hs
import Data.List(group, sort)

-- helper func to find amount of odd freq letters
countOddFreq :: [Int] -> Int
countOddFreq freqs = length (filter odd freqs) -- counts how many letters that are odd freq

-- Func that finds min nr of times to remove letters to become peragram
-- If 1 letters are oddfreq: 1-1 = 0 -> max 0 0 = 0
-- If 0 letters are oddfreq: 0-1 = -1 -> max 0 (-1) = 0
-- If n letters are oddfreq: n-1 -> max 0 n-1 = n-1
--          n-1 because ONE letter may be oddfreq at most
countRemovedLetters :: String -> Int
countRemovedLetters inputString = max 0 (countOddFreq letterCount - 1)
    where 
        letterCount = map length (group (sort inputString))

-- Main func to take terminal input and evaluate string
main :: IO()
main = do
    inputString <- getLine
    let test = countRemovedLetters inputString
    print test